LuaUI/Widgets/unit_auto_patrol_nanos.lua

```
@@ -26,61 +47,284 @@ end
+	local reclaim_energy = {CMD_RECLAIM, {x, y, z, buildDistance}, CMD_OPT_CTRL, "reclaim energy"}
+	local repair_units = {CMD_REPAIR, {x, y, z, buildDistance}, CMD_OPT_META, "repair units"}
+	local build_assist = {CMD_REPAIR, {x, y, z, buildDistance}, 0, "build assist"}
+
+	-- Patrolling doesn't do anything if you target the current location of
+	-- the unit. Point patrol towards map center.
+	local vx = mapCenterX - x
+	local vz = mapCenterZ - z
+	local patrol = {CMD_PATROL, {x + vx*25/abs(vx), y, z + vz*25/abs(vz)}, 0, "patrol"}
+
+	local commands = {}
+
+	if get_metal and use_metal and use_energy then
+		table.insert(commands, patrol)
+	else
+		if use_metal and use_energy then
+			table.insert(commands, build_assist)
+		end
+		if use_energy then
+			table.insert(commands, repair_units)
+		end
+		if get_metal and get_energy then
+			if metal > energy then
+				table.insert(commands, reclaim_energy)
+				table.insert(commands, reclaim_metal)
+			else
+				table.insert(commands, reclaim_metal)
+				table.insert(commands, reclaim_energy)
+			end
+		elseif get_metal then
+			table.insert(commands, reclaim_metal)
+		elseif get_energy then
+			table.insert(commands, reclaim_energy)
+		end
+	end
+
+	return commands
+end
+
+local function AllTrue(table)
+	for _, v in pairs(table) do
+		if not v then
+			return false
+		end
+	end
+	return true
+end
+
+local function SetupUnit(unitID)
 	local x, y, z = spGetUnitPosition(unitID)
 	if (x) then
-		-- point patrol towards map center
-		vx = mapCenterX - x
-		vz = mapCenterZ - z
-		x = x + vx*25/abs(vx)
-		z = z + vz*25/abs(vz)
+		local unitDefID = spGetUnitDefID(unitID)
+		local buildDistance = UnitDefs[unitDefID].buildDistance
+		trackedUnits[unitID] = trackedUnits[unitID] or {}
+		trackedUnits[unitID].checkTime = time + RandomInterval(checkInterval)
+		local cmds = DecideCommands(x, y, z, buildDistance)
+		TableEcho(cmds, "cmds: ")
+
+		local commandQueue = spGetCommandQueue(unitID, -1)
```

You don't need the whole command queue. Getting the command queue as a table spams temporary tables into memory, so Spring.GetUnitCurrentCommand(unitID, n) was added and applied a few years ago to cut down on table creation. You should be able to use Spring.GetUnitCurrentCommand on the first few commands to check what you need, failing early if you can.